src/libraries/System.Text.Json/Common/ReflectionExtensions.cs

```diff
@@ -229,5 +230,85 @@ public static bool IsVirtual(this PropertyInfo? propertyInfo)
             Debug.Assert(propertyInfo != null);
             return propertyInfo != null && (propertyInfo.GetMethod?.IsVirtual == true || propertyInfo.SetMethod?.IsVirtual == true);
         }
+
+        public static bool IsKeyValuePair(this Type type, Type? keyValuePairType = null)
+        {
+            if (!type.IsGenericType)
+            {
+                return false;
+            }
+
+            // Work around not being able to use typeof(KeyValuePair<,>) directly during compile-time src gen type analysis.
+            keyValuePairType ??= typeof(KeyValuePair<,>);
+
+            Type generic = type.GetGenericTypeDefinition();
+            return generic == keyValuePairType;
+        }
+
+        public static bool TryGetDeserializationConstructor(
+#if !BUILDING_SOURCE_GENERATOR
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)]
+#endif
+            this Type type,
+            bool useDefaultCtorInAnnotatedStructs,
```

Comment:
FWIW I added `useDefaultCtorInAnnotatedStructs` to the reflection-based object converter factory to cater to F# struct records, it may not be applicable in sourcegen scenaria.

Reply:
It's applicable in one case so far, `KeyValuePair<,>` deserialization. More so I'm just trying to share the implementation between source-gen & non source-gen as much as possible.